cni/cmd/istio-cni-taint/main.go

```diff
@@ -0,0 +1,188 @@
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	configOverrides := &clientcmd.ConfigOverrides{}
+	kubeConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, configOverrides)
+	config, err := kubeConfig.ClientConfig()
+	if err != nil {
+		return
+	}
+	clientset, err = client.NewForConfig(config)
+	return
+}
+
+// Log human-readable output describing the current filter and option selection
+func logCurrentOptions(ts *taint.TaintSetter, options *ControllerOptions) {
+	if options.RunAsDaemon {
+		log.Infof("Controller Option: Running as a Daemon.")
+	}
+	for _, cs := range ts.GetAllConfigs() {
+		log.Infof("ConfigSetting %s", cs.ToString())
+	}
+}
+
+//check all node, taint all unready node
+func nodeReadinessCheck(tc *taint.Controller) {
+	nodes := tc.ListAllNode()
+	for _, node := range nodes {
+		err := tc.ProcessNode(node)
+		if err != nil {
+			log.Fatalf("error: %+v in node %v", err.Error(), node.Name)
+		}
+	}
+}
+
+func main() {
+	loggingOptions.OutputPaths = []string{"stderr"}
+	loggingOptions.JSONEncoding = true
+	if err := log.Configure(loggingOptions); err != nil {
+		os.Exit(1)
+	}
+	options := parseFlags()
+
+	clientSet, err := clientSetup()
+	if err != nil {
+		log.Fatalf("Could not construct clientSet: %s", err)
+	}
+	taintSetter, err := taint.NewTaintSetter(clientSet, options.TaintOptions)
+	if err != nil {
+		log.Fatalf("Could not construct taint setter: %s", err)
+	}
+	logCurrentOptions(&taintSetter, options)
+	tc, err := taint.NewTaintSetterController(&taintSetter)
+	if err != nil {
+		log.Fatalf("Fatal error constructing taint controller: %+v", err)
+	}
+	if options.RunAsDaemon {
+		id := uuid.New().String()
+		stopCh := make(chan struct{})
+		//it will be run in leader for controller configuration and running
+		run := func(ctx context.Context) {
+			tc.Run(stopCh)
+		}
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		// listen for interrupts or the Linux SIGTERM signal and cancel
+		// our context, which the leader election code will observe and
+		// step down
+		ch := make(chan os.Signal, 1)
+		signal.Notify(ch, os.Interrupt, syscall.SIGTERM)
+		go func() {
+			<-ch
+			close(stopCh)
+			log.Info("Received termination, signaling shutdown")
+			cancel()
+		}()
+		lock := &resourcelock.LeaseLock{
+			LeaseMeta: metav1.ObjectMeta{
+				Name:      LeassLockName,
+				Namespace: LeaseLockNamespace,
+			},
+			Client: clientSet.CoordinationV1(),
+			LockConfig: resourcelock.ResourceLockConfig{
+				Identity: id,
+			},
+		}
+		//time settings are adopting default settings in client-go
+		leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+			Lock:            lock,
+			ReleaseOnCancel: true,
+			LeaseDuration:   60 * time.Second,
+			RenewDeadline:   15 * time.Second,
+			RetryPeriod:     5 * time.Second,
+			Callbacks: leaderelection.LeaderCallbacks{
+				OnStartedLeading: func(ctx context.Context) {
```

Comment:
From what I understand (and @howardjohn please correct me if I'm wrong) this callback function should be where all the run commands (such as `tc.Run`) should be executed.

See https://github.com/istio/istio/blob/022f2b525339203317075e4833e1d39bc8926841/pilot/pkg/leaderelection/leaderelection.go#L85 as a reference.